pkg/agent/openflow/network_policy.go

```diff
@@ -706,53 +708,78 @@ func (c *client) NewDNSPacketInConjunction(id uint32) error {
 	if err := c.ofEntryOperations.AddAll(conj.actionFlows); err != nil {
 		return fmt.Errorf("error when adding action flows for the DNS conjunction: %w", err)
 	}
-	udpService := v1beta2.Service{
-		Protocol: &protocolUDP,
-		SrcPort:  &dnsPort,
-	}
 	dnsPriority := priorityDNSIntercept
+	dnsCTState := &openflow15.CTStates{
+		// Use ct_state=+trk+rpl as matching condition.
+		// CTState bit-state map:
+		// dnat | snat | trk | inv | rpl | rel | est | new
+		Data: 0b00101000,
+		Mask: 0b00101000,
+	}
 	conj.serviceClause = conj.newClause(1, 2, getTableByID(conj.ruleTableID), nil)
 	conj.toClause = conj.newClause(2, 2, getTableByID(conj.ruleTableID), nil)
 	c.featureNetworkPolicy.conjMatchFlowLock.Lock()
 	defer c.featureNetworkPolicy.conjMatchFlowLock.Unlock()
-	ctxChanges := conj.serviceClause.addServiceFlows(c.featureNetworkPolicy, []v1beta2.Service{udpService}, &dnsPriority, false)
-
-	tcpFlags := TCPFlags{
-		// URG|ACK|PSH|RST|SYN|FIN|
-		Flag: 0b011000,
-		Mask: 0b011000,
-	}
-	tcpDNSPort := types.BitRange{Value: uint16(dnsPort)}
+	var ctxChanges []*conjMatchFlowContextChange
 	for _, proto := range c.featureNetworkPolicy.ipProtocols {
-		tcpServiceMatch := &conjunctiveMatch{
+		tcpMatch := &conjunctiveMatch{
+			tableID:  conj.serviceClause.ruleTable.GetID(),
+			priority: &dnsPriority,
+		}
+		udpMatch := &conjunctiveMatch{
 			tableID:  conj.serviceClause.ruleTable.GetID(),
 			priority: &dnsPriority,
 		}
 		if proto == binding.ProtocolIP {
-			tcpServiceMatch.matchPairs = []matchPair{
+			tcpMatch.matchPairs = []matchPair{
 				{
 					matchKey:   MatchTCPSrcPort,
-					matchValue: tcpDNSPort,
+					matchValue: types.BitRange{Value: uint16(dnsPort)},
+				},
+				{
```

Comment:
nit: it seems that we can start with the common
```
{
	matchKey:   MatchCTState,
	matchValue: dnsCTState,
}
```
as the matchPair and append protocol specific ones in each branch to make the logic look cleaner?

Reply:
Done.